경북대 BE_김동윤 1주차 과제(3단계)

dong-yxxn opened the pull request:
## 어려웠던점
JdbcTemplate를 이용해서 직접 db를 다루는 것이 처음이다보니 많이 어려웠습니다. 그래도 일단 부딪치면서 모르는거를 공부하다 보니 많이 알게되는것 같습니다.
아직 DatabaseInitializer가 어떻게 동작하는지 명확하지 않습니다. thymeleaf에 대해서도 아직 부족함을 느껴, 이에 대해 더 학습하도록 하곘습니다.

dong-yxxn committed:
fetch(ReadMe): 기능 요구사항 명세

dong-yxxn committed:
refactor(build.gradle): java version 21 -> 17로 변경

dong-yxxn committed:
refactor(ReadMe): Representation class 멤버 변수 타입 변경

안정성을 위해 Int -> long으로 타입 변경

dong-yxxn committed:
feat(Product): record로 클래스 생성

dong-yxxn committed:
feat(PdController): 컨트롤러 생성 및 임의 저장소 선언

dong-yxxn committed:
feat(PdController): getProduct 기능 추가
- 저장소 이름 오타 수정

dong-yxxn committed:
refactor(Product): id, price int -> long 타입으로 변경

dong-yxxn committed:
feat(PdController): CRUD 기능 구현

dong-yxxn committed:
refactor(PdController): id 입력

dong-yxxn committed:
refactor(PdController): request mapping 수정
- root를 -> 로 변경

dong-yxxn committed:
feat(ReadMe): 2단계 기능 요구사항 명세

dong-yxxn committed:
refactor(PdController): 상품목록 조회 Html 호출

dong-yxxn committed:
refactor(add_product.html): 상품추가 html 작성

dong-yxxn committed:
feat(ViewController): 상품 추가 버튼 누를 시에 'add_product.html' 사이트를 열수 있도록 표시

dong-yxxn committed:
feat(products.html): 상품 조회 html 작성

dong-yxxn committed:
refactor(PdController): POST 리디렉션 추가
- add_product.html에서 submit 누를시에, 데이터 저장 후, product.html 표시

dong-yxxn committed:
feat(products.html): 삭제 기능 추가
- 체크된 상품 삭제
- add_product.html 사소한 수정

dong-yxxn committed:
refactor(PdController): 체크된 것들 삭제 기능 구현

dong-yxxn committed:
포기

dong-yxxn committed:
refactor(ProductController): 이름 변경 및 Model 사용하도록 코드 수정

dong-yxxn committed:
refactor(ViewController): 주소 수정 및 Model 사용

dong-yxxn committed:
.

dong-yxxn committed:
refactor(products.html): thymeleaf를 좀더 횽ㄹ적으로 사용할수있도록 수정

dong-yxxn committed:
refactor(add_prodcut.html): thymeleaf를 좀 더 효율적으로 사용할수있도록 수정

dong-yxxn committed:
refactor(add_product.html): id 고정으로 인한 문제 해결
- id가 0으로 고정되어 HashMap에서 덮어써져서 저장이 안되는 오류를 직접 입력하게 만들어 고침

dong-yxxn committed:
refactor(products.html): imageUrl이 보이지 않아 text로 변경함

dong-yxxn committed:
feat(ReadMe): 3단계 기능 요구사항 명세

dong-yxxn committed:
feat(ProductRepository): JdbcTemplate 사용을 위해 Repository 클래스 선언

dong-yxxn committed:
refactor(ProductController): repository 이용하도록 코드 수정

dong-yxxn committed:
feat(DatabaseInitializer): table 생성을 위해 클래스 선성ㄴ

dong-yxxn committed:
refactor(ViewController): jdbc에 맞게 view 파트 수정

MangKyu commented on README.md:
```diff
@@ -1 +1,52 @@
-# spring-gift-product
\ No newline at end of file
+# spring-gift-product
```
요구사항 정리한 부분 좋네요 👍

MangKyu commented on src/main/java/gift/DatabaseInitializer.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class DatabaseInitializer implements CommandLineRunner {
```
애플리케이션 실행 이후 초기화를 위해 CommandLineRunner를 활용한 부분은 좋았습니다!
다만 테이블 생성과 같은 경우 설정으로 처리하는 더 좋은 방법이 있어서요! 아래 코멘트에서 자세히 남겨드리겠습니다!
해당 부분은 아래 코멘트 반영하면 삭제해도 될 것 같아요ㅎㅎ

MangKyu replied:
참고로 CommandLineRunner는 `커맨드 라인으로 받은 스트링 타입의 인자를 파라미터로 받아서 사용하기 위한 용도`이므로, ApplicationRunner를 사용하시면 됩니다!

MangKyu commented on src/main/java/gift/Product.java:
```diff
@@ -0,0 +1,5 @@
+package gift;
+
+public record Product(long id, String name, long price, String imageUrl) {
```
record는 데이터 전달(data carrier)을 위한 용도로 사용되도록 고안되었습니다!
따라서 데이터베이스를 통해 저장되고 관리되는 경우에는 record를 활용하지 않는 경우가 많습니다 🙏

MangKyu commented on src/main/java/gift/ProductController.java:
```diff
@@ -0,0 +1,41 @@
+        productRepository.save(product);
+    }
+
+    @PutMapping("/{id}")
```
적절한 HTTP Method를 사용해주셨네요! 👍

MangKyu commented on src/main/java/gift/ViewController.java:
```diff
@@ -0,0 +1,62 @@
+
+@Controller
+@RequestMapping("/admin/products")
+public class ViewController {
```
Product에 대한 view를 처리하는 부분이므로, Product도 클래스 네이밍에 포함되는 것이 좋을 것 같아요!
그렇지 않으면 다른 view 처리가 생겼을 때, 클래스 이름이 conflict 날 것 같습니다ㅎㅎ

MangKyu commented on src/main/java/gift/ProductRepository.java:
```diff
@@ -0,0 +1,50 @@
+    private final RowMapper<Product> productRowMapper = (rs, rowNum) ->
+        new Product(rs.getLong("id"), rs.getString("name"), rs.getLong("price"), rs.getString("image_url"));
+
+    public void createTable() {
```
아래 코멘트에서 참고 부탁드려요!

MangKyu commented on src/main/resources/application.properties:
```diff
@@ -1 +1,6 @@
+spring.datasource.driverClassName=org.h2.Driver
+spring.datasource.username=sa
+spring.datasource.password=password
+spring.h2.console.enabled=true
```
spring.sql.init.schema-locations=classpath:schema.sql
위와 같은 설정을 추가하면 서버가 실행될 때, 스키마를 위한 쿼리를 모아둔 sql을 실행시킬 수 있습니다!

그리고 resources 하위에 schema.sql을 추가하면 서버가 실행될 때, 스키마 설정을 보다 용이하게 관리할 수 있을 것 같습니다ㅎㅎ
추가적으로 기존 쿼리에서 not-null과 같은 제약 조건을 추가하였어요!
```java
create table `products` (
                            `id` int auto_increment primary key,
                            `name` varchar(255) not null,
                            `price` int not null,
                            `image_url` varchar(255) not null
);
```

MangKyu reviewed (changes requested):
안녕하세요 동윤님ㅎㅎ 만나서 반갑습니당 🙇
PR 올려주신 것 확인했고, 전반적으로 잘 작성해주셨는데 몇 가지 고치면 좋을 것 같아서 코멘트들 남겨두었어요ㅎㅎ
확인해보시고, 필요한 부분 반영해주시면 좋을 것 같습니당! 파이팅입니다!! 🔥🔥🔥
(앞에 1, 2단계 별도로 올려주신 PR은 여기에 한번에 달았으니 닫으셔도 될 것 같습니당 🙏 )

- step 1 
  - post나 put 요청에 대한 응답 메시지는 일반적으로 생성 및 수정으로 인한 리소스 정보를 반환합니다. 다만 이는 팀 혹은 회사마다 컨벤션을 정의하기도 합니다! 관련 내용은 [RFC7231](https://datatracker.ietf.org/doc/html/rfc7231#section-6.3.1)이라는 API 규격 문서를 보시면 좋을 것 같습니다ㅎㅎ github과 같은 서비스들의 open api(공개된 api)를 통해 사례를 파악하는 것도 좋은 학습 방법입니다!
- step 2
  - thymeleaf의 경우 저도 처음 학습하는데 어려웠던 기억이 나네요ㅎㅎ 최근에는 화면 부분은 FE 개발자를 통해서 처리하는 추세이니, 그 외에 나머지를 먼저 학습하시면 될 것 같습니다!
- step 3
  - 스프링 프레임워크는 요청을 받을 준비가 되면, 마지막으로 필요한 정보들을 초기화를 위한 단계에 진입합니다. 이때 CommandLineRunner 또는 ApplicationRunner들을 일괄적으로 실행시킵니다!

dong-yxxn committed:
refactor: controller, model, repository 패키지 생성

dong-yxxn committed:
refactor(PVController): 이름 충돌 방지를 위해 변경

dong-yxxn committed:
feat(schma.sql): 테이블 생성 방식을 schema.sql로 변경함
- DatabaseInitializer.java -> schema.sql 이용하도록 properties 파일 수정

dong-yxxn committed:
refactor: Product를 record -> public으로 변환
- ProductRepository, add_product.html, products.html를 이에 맞게 수정
- public 변환 후, getter 메소드 정의

dong-yxxn committed:
refactor(add_product.html): id창 입력 비활성화
- slq에서 id가 auto_increased로 설정되어, id 값 입력이 의미없다 생각되어 readonly로 변경함

dong-yxxn committed:
refactor(ProductController): ResponseEntity 사용하여 상태코드 반환

dong-yxxn committed:
refactor(ProductViewController): ProductController 변경에 대응하여 코드 변경
- .getBody() 메소드 부분적으로 추가
